### PR TITLE
HBAVSS: implement implicate handling and share recovery

### PR DESCRIPTION
**Pull request by nqyy**
The implementation of implicate handling part and share recovery part is complete. They are tested separately in the code. The whole structure of the code is changed and it doesn't return a value but set a future value for output instead, as you can see in the code change of the test.

A new method of interpolation on G1 is added to betterpairing.

New tests are also added.

### CODE REVIEW

File: honeybadgermpc/hbavss.py
Comment (nqyy) [inline honeybadgermpc/hbavss.py]: @tyurek said he'll implement the interpolation on the commitments. So, I commented this part out for now.

**Comment by tyurek** [comment]
I'm thinking we wait until we have implicate fully working before trying to merge this

**Comment by amiller** [comment]
@tyurek said he will prefer to review this after the new polycommit is integrated

**Comment by nqyy** [comment]
Will update this once https://github.com/initc3/HoneyBadgerMPC/pull/257 is merged.

**Comment by nqyy** [comment]
Fully updated.

**Comment by amiller** [comment]
Tom and I discussed and realized that something this is missing from this for our use case is using a discrete log proofs about the encryption key, rather than the key itself. We know what the next step is for that, but I think it can go as a separate PR.

File: honeybadgermpc/betterpairing.py
Comment (smkuls) [inline honeybadgermpc/betterpairing.py:723]: I am not okay with this change. We have a whole module dedicated to polynomial operations - `polynomial.py`. We should try to make use of that instead of duplicating functionality just because we are operating on a different type.

File: honeybadgermpc/hbavss.py
Comment (smkuls) [inline honeybadgermpc/hbavss.py:308]: Consider splitting this method, this is an extremely long method.

File: honeybadgermpc/hbavss.py
Comment (smkuls) [inline honeybadgermpc/hbavss.py:352]: what's `r1_set`? I am not aware of the protocol but can you pick a more descriptive name?

File: honeybadgermpc/hbavss.py
Comment (smkuls) [inline honeybadgermpc/hbavss.py]: nit:
`read_sent = r1_sent = r2_sent = False`

File: honeybadgermpc/hbavss.py
Comment (smkuls) [inline honeybadgermpc/hbavss.py]: `logger.warn`

File: honeybadgermpc/hbavss.py
Comment (smkuls) [inline honeybadgermpc/hbavss.py]: calling `encode()` is faster than calling `encode("utf-8")`

File: honeybadgermpc/hbavss.py
Comment (smkuls) [inline honeybadgermpc/hbavss.py:388]: This should not be using lagrange - we should switch to Vandermonde/FFT. I think we got rid of lagrange from most places in our code.

File: honeybadgermpc/hbavss.py
Comment (smkuls) [inline honeybadgermpc/hbavss.py]: same here, this looks like you are interpolating at all `n` points - consider using Vandermonde/FFT

File: tests/test_hbavss.py
Comment (smkuls) [inline tests/test_hbavss.py]: use `asyncio.gather` this serializes the code.

File: tests/test_hbavss.py
Comment (smkuls) [inline tests/test_hbavss.py]: nit:
`hbavss_list = [None]*n`

File: honeybadgermpc/hbavss.py
Comment (smkuls) [inline honeybadgermpc/hbavss.py:512]: add a log statement before and after, this helps in debugging what got stuck where

File: honeybadgermpc/hbavss.py
Comment (smkuls) [inline honeybadgermpc/hbavss.py:474]: Can we avoid a double `dumps` here? Can you not pickle them just here and get rid of the previous pickle?

File: honeybadgermpc/betterpairing.py
Comment (smkuls) [reply honeybadgermpc/betterpairing.py:723]: That module actually caches interpolations and here you don't get that benefit. I would strongly suggest using that code since that contains similar operations in one file making the code more maintainable in the future.

**Review by smkuls — CHANGES_REQUESTED** [review]

File: honeybadgermpc/betterpairing.py
Comment (nqyy) [reply honeybadgermpc/betterpairing.py:723]: @tyurek

File: honeybadgermpc/hbavss.py
Comment (nqyy) [reply honeybadgermpc/hbavss.py:352]: That message is recovery phase 1 and it is called exactly as R1 in the paper

**Comment by smkuls** [comment]
@chitianhao I don't see any tests as well. You should cover the entire code - this includes the implicate and recovery phase. We never got to adding proper tests for your last PR so I'd want to have them checked in along with this.

File: honeybadgermpc/hbavss.py
Comment (nqyy) [reply honeybadgermpc/hbavss.py]: Yeah, but we also encode it this way in the light version.

Comment (smkuls) [reply honeybadgermpc/hbavss.py]: You can change it both places. :)

File: tests/test_hbavss.py
Comment (nqyy) [reply tests/test_hbavss.py]: The batch avss doesn't return anything now. It output the value in a Future variable.

File: honeybadgermpc/hbavss.py
Comment (nqyy) [reply honeybadgermpc/hbavss.py:474]: No, they are separate and one is for RBC and the other one (need to be a list) for AVID.

**Comment by nqyy** [comment]
Yeah, I think we should add test for that. @tyurek is trying to find a way to test this part properly.

File: honeybadgermpc/hbavss.py
Comment (nqyy) [reply honeybadgermpc/hbavss.py:308]: Yep. I'll see what I can do.

File: honeybadgermpc/hbavss.py
Comment (nqyy) [reply honeybadgermpc/hbavss.py:388]: @tyurek

File: tests/test_hbavss.py
Comment (smkuls) [reply tests/test_hbavss.py]: You can collect all futures and then do a gather on them:
```
shares = await asyncio.gather(*[hbavss_list[i].shares_future for i in range(n)])
```

Comment (nqyy) [reply tests/test_hbavss.py]: Yeah, this is a good way.

File: honeybadgermpc/hbavss.py
Comment (tyurek) [reply honeybadgermpc/hbavss.py:388]: I don't have code handy to do FFT interpolation for commitments in G1. If someone wants to tackle it, feel free, but I don't think it's terribly important as this is just fault handling anyway. It would be new work with little benefit

File: honeybadgermpc/hbavss.py
Comment (tyurek) [reply honeybadgermpc/hbavss.py]: As this is ZR, FFT should be used, yeah

**Review by tyurek — CHANGES_REQUESTED** [review]
Implicate handling for hbavss-lite would be a good thing to include. Needs tests as well

File: honeybadgermpc/hbavss.py
Comment (tyurek) [reply honeybadgermpc/hbavss.py:388]: interpolation in a finite field should use the polynomial class and FFT and all that though

**Comment by nqyy** [comment]
Thanks to @tyurek for adding the test. And I also fixed an issue with it,

**Review by tyurek — APPROVED** [review]

File: honeybadgermpc/hbavss.py
Comment (Drake-Eidukas) [inline honeybadgermpc/hbavss.py]: Is there ever a case where both of these `if` statements are supposed to be true? If not, we should make this second `if` and `elif` to make this clearer.

File: honeybadgermpc/hbavss.py
Comment (Drake-Eidukas) [inline honeybadgermpc/hbavss.py]:
```suggestion
return not self.poly_commit.verify_eval(
    commitments[j_k], j+1, j_share, j_aux, j_witnesses)
```

File: honeybadgermpc/hbavss.py
Comment (Drake-Eidukas) [inline honeybadgermpc/hbavss.py]: Same as other comment

File: tests/test_hbavss.py
Comment (Drake-Eidukas) [inline tests/test_hbavss.py:318]: This looks copy-pasted from the other tests-- if so, could you move it into `tests/fixtures`? It'll be better for future maintenance if this is just defined once.

**Review by Drake-Eidukas — CHANGES_REQUESTED** [review]
Mostly style nits. Could you also squash the commits and make a unified commit message?

File: tests/test_hbavss.py
Comment (tyurek) [reply tests/test_hbavss.py:318]: it's slightly different each time. The tests have the dealer introduce faults in differing ways

Comment (tyurek) [reply tests/test_hbavss.py:318]: repeatedly overriding the same class function was just the most convenient way to do that

Comment (Drake-Eidukas) [reply tests/test_hbavss.py:318]: Fair enough-- they looked _mostly_ the same. It might be a good idea to leave a comment describing how they introduce faults to help anyone that has to work on this in the future then.

Comment (tyurek) [reply tests/test_hbavss.py]: unless you meant the callback?

File: honeybadgermpc/betterpairing.py
Comment (Drake-Eidukas) [inline honeybadgermpc/betterpairing.py:743]: This will be faster if you do--
```
sorted_coords = sorted(coords, key=lambda x: x[0])
xs = [coord[0] for coord in sorted_coords]
```

File: tests/test_hbavss.py
Comment (Drake-Eidukas) [inline tests/test_hbavss.py:99]: This part with ExitStack() is definitely duplicated across tests.

**Review by Drake-Eidukas — APPROVED** [review]
I would like it if you can pull out the stuff using `with ExitStack() as stack` into a method that you call rather than have it duplicated, but other than that, this looks good to me. I'll wait on a final review from @smkuls

File: tests/test_hbavss.py
Comment (smkuls) [reply tests/test_hbavss.py:99]: Lets take this later. This PR is too blown up. Lets not take up any more changes in this one. This is blocking other people as well.

File: honeybadgermpc/betterpairing.py
Comment (smkuls) [reply honeybadgermpc/betterpairing.py:723]: @tyurek are these methods still needed?

Comment (smkuls) [reply honeybadgermpc/betterpairing.py:723]: I thought these were moved these to the polynomial class?

**Review by smkuls — COMMENTED** [review]
As of this moment there's way too much code for this to be reviewed properly.

From next time, lets split it in multiple PRs. As a feedback, you can always reply to a review suggestion that you want to take something later and immediately follow up with another PR. 

I am approving this.

**Review by smkuls — APPROVED** [review]

**Comment by nqyy** [comment]
Rebased. Let's ship it.

**Pull request merged**
